src/renderers/widget.ts

```diff
@@ -0,0 +1,350 @@
+
+import {
+  Message
+} from 'phosphor/lib/core/messaging';
+
+import {
+  Widget
+} from 'phosphor/lib/ui/widget';
+
+import {
+  typeset, removeMath, replaceMath
+} from './latex';
+
+
+/*
+ * The class name added to rendered HTML. 
+ */
+const HTML_CLASS = 'jp-RenderedHTML';
+
+/*
+ * The class name added to rendered markdown. 
+ */
+const MARKDOWN_CLASS = 'jp-RenderedMarkdown';
+
+/*
+ * The class name added to rendered Latex. 
+ */
+const LATEX_CLASS = 'jp-RenderedLatex';
+
+/*
+ * The class name added to rendered images. 
+ */
+const IMAGE_CLASS = 'jp-RenderedImage';
+
+/*
+ * The class name added to rendered text. 
+ */
+const TEXT_CLASS = 'jp-RenderedText';
+
+/*
+ * The class name added to rendered javascript. 
+ */
+const JAVASCRIPT_CLASS = 'jp-RenderedJavascript';
+
+/*
+ * The class name added to rendered SVG. 
+ */
+const SVG_CLASS = 'jp-RenderedSVG';
+
+/*
+ * The class name added to rendered PDF. 
+ */
+const PDF_CLASS = 'jp-RenderedPDF';
+
+
+// Support GitHub flavored Markdown, leave sanitizing to external library.
+marked.setOptions({
+  gfm: true,
+  sanitize: false,
+  breaks: true,
+  langPrefix: 'cm-s-default language-',
+  highlight: (code, lang, callback) => {
+    if (!lang) {
+        // no language, no highlight
+        if (callback) {
+            callback(null, code);
+            return;
+        } else {
+            return code;
+        }
+    }
+    requireMode(lang).then(spec => {
+      let el = document.createElement('div');
+      if (!spec) {
+          console.log(`No CodeMirror mode: ${lang}`);
+          callback(null, code);
+          return;
+      }
+      try {
+        CodeMirror.runMode(code, spec.mime, el);
+        callback(null, el.innerHTML);
+      } catch (err) {
+        console.log(`Failed to highlight ${lang} code`, err);
+        callback(err, code);
+      }
+    }).catch(err => {
+      console.log(`No CodeMirror mode: ${lang}`);
+      console.log(`Require CodeMirror mode error: ${err}`);
+      callback(null, code);
+    });
+  }
+});
+
+
+/**
+ * A widget for displaying HTML and rendering math.
+ */
+export
+class RenderedHTML extends Widget {
+  /**
+   * Construct a new html widget.
+   */
+  constructor(options: RenderMime.IRenderOptions) {
+    super();
+    this.addClass(HTML_CLASS);
+    let source = options.source;
+    if (options.sanitizer) {
+      source = options.sanitizer.sanitize(source);
+    }
+    appendHtml(this.node, source);
+    if (options.resolver) {
+      resolveUrls(this.node, options.resolver);
+    }
+  }
+
+  /**
+   * A message handler invoked on an `'after-attach'` message.
+   */
+  onAfterAttach(msg: Message): void {
+    typeset(this.node);
+  }
+}
+
+
+/**
+ * A widget for displaying Markdown.
+ */
+export
+class RenderedMarkdown extends Widget {
+  /**
+   * Construct a new markdown widget.
+   */
+  constructor(options: RenderMime.IRenderOptions) {
+    super();
+    this.addClass(MARKDOWN_CLASS);
+    this.addClass(HTML_CLASS);
```

Comment:
I don't agree with this. `RenderedMarkdown` does not inherit from `RenderedHTML`, so it shouldn't have that CSS class. Maybe we need a token base class which does nothing except add a common CSS class, which is inherited by every widget in this file..